impacket/ldap/ldapasn1.py

```diff
@@ -14,209 +14,52 @@
+    'BindResponse', 'UnbindRequest', 'SubstringFilter', 'MatchingRuleAssertion', 'Filter', 'SearchRequest',
+    'SearchResultEntry', 'SearchResultReference', 'SearchResultDone', 'ModifyRequest', 'ModifyResponse', 'AddRequest',
+    'AddResponse', 'DelRequest', 'DelResponse', 'ModifyDNRequest', 'ModifyDNResponse', 'CompareRequest',
+    'CompareResponse', 'AbandonRequest', 'ExtendedRequest', 'ExtendedResponse', 'IntermediateResponse', 'Control',
+    'Controls', 'SimplePagedResultsControlValue', 'SimplePagedResultsControl', 'LDAPMessage'
+]
+
+# Controls
 CONTROL_PAGEDRESULTS = '1.2.840.113556.1.4.319'
 
 KNOWN_CONTROLS = {}
 
-# LDAP Unsolicited Notifications
+# Unsolicited notifications
 NOTIFICATION_DISCONNECT = '1.3.6.1.4.1.1466.20036'
 
 KNOWN_NOTIFICATIONS = {NOTIFICATION_DISCONNECT: 'Notice of Disconnection'}
 
-
-################################################################################
-# CLASSES
-################################################################################
-
-class Integer7Bit(Integer):
-    # INTEGER (1 ..  127)
-    subtypeSpec = ValueRangeConstraint(1, 127)
-
-
-class IntegerPositive(Integer):
-    subtypeSpec = Integer.subtypeSpec + ValueRangeConstraint(0, MAXINT)
-
-
-class LDAPString(OctetString):
-    """
-        LDAPString ::= OCTET STRING -- UTF-8 encoded,
-                                    -- [ISO10646] characters
-    """
-    encoding = 'utf-8'
-    pass
-
-
-class LDAPDN(LDAPString):
-    """
-        LDAPDN ::= LDAPString -- Constrained to <distinguishedName>
-                              -- [RFC4514]
-    """
-    pass
+maxInt = univ.Integer(2147483647)
 
 
-class AuthSimple(OctetString):
-    tagSet = OctetString.tagSet.tagImplicitly(Tag(tagClassContext, tagFormatSimple, 0))
-    encoding = 'utf-8'
-
-
-class Credentials(OctetString):
-    encoding = 'utf-8'
-
-
-class SaslCredentials(Sequence):
-    """
-        SaslCredentials ::= SEQUENCE {
-             mechanism               LDAPString,
-             credentials             OCTET STRING OPTIONAL }
-    """
-    tagSet = Sequence.tagSet.tagImplicitly(Tag(tagClassContext, tagFormatConstructed, 3))
-    componentType = NamedTypes(
-        NamedType('mechanism', LDAPString()),
-        OptionalNamedType('credentials', Credentials())
-    )
-
-
-class SicilyPackageDiscovery(OctetString):
-    tagSet = OctetString.tagSet.tagImplicitly(Tag(tagClassContext, tagFormatSimple, 9))
-    encoding = 'utf-8'
-
-
-class SicilyNegotiate(OctetString):
-    tagSet = OctetString.tagSet.tagImplicitly(Tag(tagClassContext, tagFormatSimple, 10))
-    encoding = 'utf-8'
+class DefaultSequenceAndSetBaseMixin:
+    def getComponentByPosition(self, idx):
+        for cls in self.__class__.__bases__:
+            if cls is not DefaultSequenceAndSetBaseMixin:
+                try:
+                    component = cls.getComponentByPosition(self, idx)
+                except AttributeError:
+                    continue
+                if component is None:
```

Comment:
I think not. The only exception we want to catch is `AttributeError` in case `cls` has no `getComponentByPosition` method, in which case the loop will `continue` to the next base class. Any other error will propagate up.

I noticed that this approach has one drawback. Mainly when there is an uninitialized component you get this error:
```'No value for %s()' % attr```
instead of this:
``'Uninitialized component #%s at %r' % (idx, self)```

Do you think we can live with it? It's not super cryptic and I think the ease of use makes up for it.